api/pkg/transformer/feast/transformer.go

```diff
@@ -181,31 +208,113 @@ func (t *Transformer) getFeastFeature(ctx context.Context, tableName string, req
 		features = append(features, feature.Name)
 	}
 
-	feastRequest := feast.OnlineFeaturesRequest{
-		Project:  config.Project,
-		Entities: entities,
-		Features: features,
-	}
-	t.logger.Debug("feast_request", zap.Any("feast_request", feastRequest))
-
-	startTime := time.Now()
-	feastResponse, err := t.feastClient.GetOnlineFeatures(ctx, &feastRequest)
-	durationMs := time.Now().Sub(startTime).Milliseconds()
+	feastFeature, err := t.feastFeatureBatchFetching(ctx, entities, config, features)
 	if err != nil {
-		feastLatency.WithLabelValues("error").Observe(float64(durationMs))
-		feastError.Inc()
 		return nil, err
 	}
-	feastLatency.WithLabelValues("success").Observe(float64(durationMs))
+	return feastFeature, nil
+}
 
-	t.logger.Debug("feast_response", zap.Any("feast_response", feastResponse.Rows()))
+type batchResult struct {
+	featuresData FeaturesData
+	err          error
+}
 
-	feastFeature, err := t.buildFeastFeatures(ctx, feastResponse, config)
-	if err != nil {
-		return nil, err
+func (t *Transformer) feastFeatureBatchFetching(ctx context.Context, entities []feast.Row, config *transformer.FeatureTable, features []string) (*FeastFeature, error) {
+
+	var cachedValues FeaturesData
+	entityNotInCache := entities
+
+	if t.options.CacheEnabled {
+		cachedValues, entityNotInCache = fetchFeaturesFromCache(t.cache, entities)
 	}
 
-	return feastFeature, nil
+	numOfBatchBeforeCeil := float64(len(entityNotInCache)) / float64(t.options.BatchSize)
+	numOfBatch := int(math.Ceil(numOfBatchBeforeCeil))
+
+	batchResultChan := make(chan batchResult, numOfBatch)
+	feastColumns := t.getFeastColumns(config)
+	entityIndices := t.getEntityIndicesFromColumns(feastColumns, config.Entities)
+	for i := 0; i < numOfBatch; i++ {
+		startIndex := i * t.options.BatchSize
+		batchedEntities := entityNotInCache[startIndex:]
+		if i != numOfBatch-1 {
```

Comment:
It can be implified as
```
		startIndex := i * t.options.BatchSize
		endIndex := min(len(entityNotInCache) - 1, startIndex+t.options.BatchSize)
		batchedEntities := entityNotInCache[startIndex:endIndex]
```